### PR TITLE
Some casemapping tests, armenian ligature support, doctests

### PR DESCRIPTION
**Pull request by Manishearth**
Progress on #3234

Not sure I'm happy with test coverage, would like a second opinion.

### CODE REVIEW

File: experimental/casemapping/src/casemapping.rs
Comment (robertbastian) [inline experimental/casemapping/src/casemapping.rs:41]: The boilerplate should be at the end of the docs

File: experimental/casemapping/src/casemapping.rs
Comment (robertbastian) [inline experimental/casemapping/src/casemapping.rs:62]: thought: this wouldn't be hard to add, we can just go through the fallback chain until something non-rooty appears. Actually it's weird that this uses `&Locale`, all other constructors use `&DataLocale`.

**Review by robertbastian — APPROVED** [review]

File: experimental/casemapping/src/casemapping.rs
Comment (eggrobin) [inline experimental/casemapping/src/casemapping.rs:197]:
```suggestion
    /// Returns the simple case folding of the given char.
```
We have mappings and foldings, calling the folding a mapping is going to add to already widespread confusion.

File: experimental/casemapping/src/casemapping.rs
Comment (eggrobin) [inline experimental/casemapping/src/casemapping.rs:200]: I think the statement being made here is that two single-character strings are caseless matches (as defined by D44 of TUS) if and only if the result of `fold` is the same for the two characters; in other words:
```rs
(cm.fold(x) == cm.fold(y)) ==
(cm.full_fold_string(x.to_string()) == cm.full_fold_string(y.to_string()))
```

**However:** This is not true until Unicode Version 15.1, see [UTC-175-C19](https://www.unicode.org/L2/L2023/23076.htm#175-C19).

Counterexample before 15.1:
```rs
let x = 'ﬅ';
let y = 'ﬆ';
```

Comment (eggrobin) [inline experimental/casemapping/src/casemapping.rs:200]: Extreme editorial nitpickery:
```suggestion
    /// You can use the case folding to compare characters case-insensitively.
```

File: experimental/casemapping/src/casemapping.rs
Comment (eggrobin) [inline experimental/casemapping/src/casemapping.rs:209]:
```suggestion
    /// // Perform caseless matches:
```

File: experimental/casemapping/src/casemapping.rs
Comment (eggrobin) [inline experimental/casemapping/src/casemapping.rs:273]:
```suggestion
    /// Can be used to test if two strings are case-insensitively equivalent.
    /// Specifically, this function implements the operation [toCasefold](https://www.unicode.org/versions/Unicode15.0.0/ch03.pdf#G53253), and can be
    // used to implement [Default Caseless Matching](https://www.unicode.org/versions/latest/ch03.pdf#G34145).
```
It is important to be specific here, because there is also canonical caseless match, compatibility caseless match, and identifier caseless match, all of which are some flavour of case-insensitive equivalence.

File: experimental/casemapping/src/casemapping.rs
Comment (eggrobin) [inline experimental/casemapping/src/casemapping.rs:350]:
```suggestion
    /// Can be used to test if two strings are case-insensitively equivalent.
    /// Specifically, this function implements the operation [toCasefold](https://www.unicode.org/versions/Unicode15.0.0/ch03.pdf#G53253), and can be
    // used to implement [Default Caseless Matching](https://www.unicode.org/versions/latest/ch03.pdf#G34145).
```

**Review by eggrobin — CHANGES_REQUESTED** [review]

File: experimental/casemapping/src/casemapping.rs
Comment (Manishearth) [reply experimental/casemapping/src/casemapping.rs:62]: ICU4C does this as well and it makes sense: you kinda want exact matches here as long as it's documented clearly. The region should not affect the casemapping algorithm, the language should. 

We're not loading data here based on this, so it shouldn't be a data locale. Could be a language though.

Comment (robertbastian) [reply experimental/casemapping/src/casemapping.rs:62]: > We're not loading data here based on this, so it shouldn't be a data locale.

This is an implementation detail.

File: experimental/casemapping/src/casemapping.rs
Comment (eggrobin) [inline experimental/casemapping/src/casemapping.rs:365]:
```suggestion
    /// // Check if two strings are caseless matches:
```

File: experimental/casemapping/src/casemapping.rs
Comment (Manishearth) [reply experimental/casemapping/src/casemapping.rs]: I can tweak this to say "provided this case folds to a single codepoint" since there's already a caveat about this

File: experimental/casemapping/src/casemapping.rs
Comment (eggrobin) [inline experimental/casemapping/src/casemapping.rs:396]:
```suggestion
    /// // Check if two strings case-insensitively equivalent under Turkic rules:
```

File: experimental/casemapping/src/casemapping.rs
Comment (Manishearth) [reply experimental/casemapping/src/casemapping.rs:62]: I don't really think it is; it's a different kind of locale parameter in my book.

File: experimental/casemapping/src/casemapping.rs
Comment (Manishearth) [reply experimental/casemapping/src/casemapping.rs]: How does this look?

File: experimental/casemapping/src/casemapping.rs
Comment (robertbastian) [reply experimental/casemapping/src/casemapping.rs:62]: Alternatively, we could move the parameter to the methods, as it's not needed in the constructor. There are already special `_turkic` methods, which we could replace by more general `_for_locale` methods.

Comment (Manishearth) [reply experimental/casemapping/src/casemapping.rs:62]: I'm fine with that. I don't think it needs to be on the constructor and was considering bringing it up as a part of final API polish (I've already listed a bunch of other things on the related issue)

We can expose the enum and also a From/Default.

How about we merge this as-is and I do the rest in a separate PR

Comment (robertbastian) [reply experimental/casemapping/src/casemapping.rs:62]: Sure, this is not blocking.

The `Greek` and `Dutch` variants are actually unused.

Comment (Manishearth) [reply experimental/casemapping/src/casemapping.rs:62]: Yeah there's some Greek special casing that we don't yet support

Comment (eggrobin) [reply experimental/casemapping/src/casemapping.rs:62]: > There are already special `_turkic` methods, which we could replace by more general `_for_locale` methods.

We should really not do that. The Turkic case _foldings_ need to be handled with extreme care, and must not be confused with the actual locale-dependence of case _mappings_.

See Markus’s comments on https://github.com/unicode-org/icu4x/issues/3151#issuecomment-1452524811.

File: experimental/casemapping/src/casemapping.rs
Comment (eggrobin) [inline experimental/casemapping/src/casemapping.rs:202]: I am not sure this is understandable or actionable by a normal human being.

Since we added an invariant test for character equivalence I am comfortable with stating it, just noting that there are edge cases in 15.0.

I would also mention the main real application of this function, which is simple case folding, wherever required.

```suggestion
    /// This function can be used to perform caseless matches on
    /// individual characters.
    /// > *Note:* With Unicode 15.0 data, there are three
    /// > pairs of characters for which equivalence under this
    /// > function is inconsistent with equivalence of the
    /// > one-character strings under [`CaseMapping::full_fold`].
    /// > This is resolved in Unicode 15.1 and later.
    ///
    /// For compatibility applications where simple case folding
    /// of strings is required, this function can be applied to
    /// each character of a string.  Note that the resulting
    /// equivalence relation is different from that obtained
    /// by [`CaseMapping::full_fold`]:
    /// The strings "Straße" and "STRASSE" are distinct
    /// under simple case folding, but are equivalent under
    /// default (full) case folding.
```

File: experimental/casemapping/src/casemapping.rs
Comment (eggrobin) [reply experimental/casemapping/src/casemapping.rs]: (Replied in a more up-to-date diff.)

File: experimental/casemapping/src/casemapping.rs
Comment (Manishearth) [reply experimental/casemapping/src/casemapping.rs]: Thanks, that seems much better

File: experimental/casemapping/src/casemapping.rs
Comment (robertbastian) [reply experimental/casemapping/src/casemapping.rs:62]: Ah I hadn't seen that. Continuing the discussion on #3234

**Comment by robertbastian** [comment]
LGTM, I'll let @eggrobin approve once he's happy

**Review by eggrobin — APPROVED** [review]

**Pull request merged**
